packages/protocol/contracts/tokenvault/erc20/registry/ERC20NativeRegistry.sol

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.20;
+
+import "../../../common/EssentialContract.sol";
+import "./IERC20NativeRegistry.sol";
+import "../adapters/BaseAdapter.sol";
+import "../../../libs/LibDeploy.sol";
+
+interface IRemoveMapping {
+    function resetCanonicalToBridged(uint256 chainId, address canonicalAddrToReset) external;
+}
+
+/// @title ERC20NativeRegistry
+/// A registry that helps with custom erc20 bridging based on some standards (e.g.: circle's USDC).
+/// It keeps track of mapping between native and counterpart contracts.
+contract ERC20NativeRegistry is EssentialContract, IERC20NativeRegistry {
+    struct PredeployedToCanonicalData {
+        address canonicalERC20;
+        address adapterAddress;
+    }
+
+    // This is needed for bridging native to CHAIN_X (_getOrDeployBridgedToken() on L2 if this is
+    // the USDC)
+    mapping(address => address) public canonicalToPredeployed;
+    // This is needed for bridging native back to NATIVE_CHAIN (_handleMessage() on L2 if this is
+    // the USDC)
+    mapping(address l2Address => PredeployedToCanonicalData) public predeployedToCanonical;
+
+    uint256[48] private __gap;
+
+    event Erc20CustomAdded(address l1Address, address l2Counterpart);
+    event Erc20CustomDeleted(address l1Address, address l2Counterpart);
+
+    error ERC20_HOOK_INVALID_ADDRESS();
+    error ERC20_HOOK_INVALID_BURN_SIGNATURE_ID();
+
+    function init(address _addressManager) external initializer {
+        EssentialContract._Essential_init(_addressManager);
+    }
+
+    /// @notice Adds a custom, pre-deployed contract mapping to it's L1 (parent chain) counterpart
+    /// @param l1Address The address of the token on L1 (or parent chain)
+    /// @param deployedCounterpart The address on L2 - it can be address(0) in which case we delete
+    /// the mappping (see explanation below).
+    /// @param adapterName The name based on which the adapter is queriable via the
+    /// AddressManager
+    /// @param deleteCustom Indicates deleting both of the mappings
+    /// @param chainId ChainId the canonical originates from
+    function changeCustomToken(
+        address l1Address,
+        address deployedCounterpart,
+        bytes32 adapterName,
+        uint64 chainId,
+        bool deleteCustom
+    )
```

Comment:
Feels a bit strange to have a single function with a `deleteCustom` variable while it doesn't really share any code. Why not create two functions, one for adding and one for removing?